<h1><center><font color='blue'>Regresión Lineal en 1D: Predicción</h1></font></center>


<h2>Objetivo</h2><ul><li> Realizar predicciones para múltiples entradas</li><li> Usar la clase Linear para construir modelos complejos</li><li> Construir un módulo personalizado</li></ul> 


<h2>Tabla de contenido</h2>
<ul>
    <li><a href="#Prediction">Predicción</a></li>
    <li><a href="#Linear">Clase Linear</a></li>
    <li><a href="#Cust">Construir módulos personalizados</a></li>
</ul>

<hr>


<h2>Preparación</h2>


Importamos las librerías necesarias:


In [23]:
# These are the libraries will be used for this lab.

import torch

<!--Empty Space for separating topics-->


<h2 id="Prediction">Predicción</h2>


Crearemos las siguientes expresiones:


$b=-1,w=2$

$\hat{y}=-1+2x$


Primero definimos los parámetros:


In [24]:
# definimos w = 2 y b = -1 para y = wx + b

w = torch.tensor(2.0, requires_grad = True)
b = torch.tensor(-1.0, requires_grad = True)

Definimos la función <code>forward(x, w, b)</code> que realiza la predicción: 


In [25]:
# Function forward(x) for prediction

def forward(x):
    yhat = w * x + b
    return yhat

Realicemos la siguiente predicción para <i>x = 1</i>


$\hat{y}=-1+2x$

$\hat{y}=-1+2(1)$


In [26]:
# predecimos y = 2x - 1 en x = 1

x = torch.tensor([[1.0]])
yhat = forward(x)
print("The prediction: ", yhat)

The prediction:  tensor([[1.]], grad_fn=<AddBackward0>)


<!--Empty Space for separating topics-->


Ahora hagamos la predicción para múltiples entradas:


<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2.png" width="500" alt="Linear Regression Multiple Input Samples" />


Primero construyamos el tensor x y chequemaos su shape:


In [27]:
# Create x Tensor and check the shape of x tensor

x = torch.tensor([[1.0], [2.0]])
print("The shape of x: ", x.shape)

The shape of x:  torch.Size([2, 1])


Ahora realizamos la predicción:


In [28]:
# Make the prediction of y = 2x - 1 at x = [1, 2]

yhat = forward(x)
print("The prediction: ", yhat)

The prediction:  tensor([[1.],
        [3.]], grad_fn=<AddBackward0>)


<!--Empty Space for separating topics-->


<h3>Práctica</h3>


Realice una predicción para el x siguiente:


In [29]:
# Practice: Make a prediction of y = 2x - 1 at x = [[1.0], [2.0], [3.0]]

x = torch.tensor([[1.0], [2.0], [3.0]])
yhat = forward(x)
print("The prediction: ", yhat)

The prediction:  tensor([[1.],
        [3.],
        [5.]], grad_fn=<AddBackward0>)


<!--Empty Space for separating topics-->


<h2 id="Linear">Clase Linear</h2>


La clase Linear puede ser utilizada para realizar una predicción. También podemos usarla para construir modelos complejos. Primero importemos el módulo:


In [30]:
# Import Class Linear

from torch.nn import Linear

Establecemos la semilla aleatoria ya que los parámetros son inicializados aleatoriamente:


In [31]:
# Set random seed

torch.manual_seed(1)

<!--Empty Space for separating topics-->


Creamos el objeto linear usando el constructor. Los parámetros son creados aleatoriamente. Imprimamos para ver w y b. Los parámetros de un modelo <code>torch.nn.Module</code> están contenidos en los parámetros del modelo y son accesados vía <code>lr.parameters()</code>:



In [32]:
# Create Linear Regression Model, and print out the parameters

lr = Linear(in_features=1, out_features=1, bias=True)
print("Parameters w and b: ", list(lr.parameters()))


Parameters w and b:  [Parameter containing:
tensor([[0.5153]], requires_grad=True), Parameter containing:
tensor([-0.4414], requires_grad=True)]


Esto es equivalente a la siguiente expresión:


$b=-0.44, w=0.5153$

$\hat{y}=-0.44+0.5153x$


El método <code>state_dict()</code> retorna un diccionario correspondiente a las capas de cada parámetro tensor.



In [33]:
print("Python dictionary: ",lr.state_dict())
print("keys: ",lr.state_dict().keys())
print("values: ",lr.state_dict().values())

Python dictionary:  OrderedDict([('weight', tensor([[0.5153]])), ('bias', tensor([-0.4414]))])
keys:  odict_keys(['weight', 'bias'])
values:  odict_values([tensor([[0.5153]]), tensor([-0.4414])])


Las claves corresponden al nombre de los atributos y los valores se obtienen vía values.


In [34]:
print("weight:",lr.weight)
print("bias:",lr.bias)

weight: Parameter containing:
tensor([[0.5153]], requires_grad=True)
bias: Parameter containing:
tensor([-0.4414], requires_grad=True)


Realicemos una predicción en <i>x = [[1.0]]</i>.


In [35]:
# Make the prediction at x = [[1.0]]

x = torch.tensor([[1.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739]], grad_fn=<AddmmBackward>)


<!--Empty Space for separating topics-->


Pueden realizarse múltiples predicciones:


<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2vector_function.png" width="500" alt="Linear Class Sample with Multiple Inputs" />


Use <code>lr(x)</code> para predecir el resultado


In [36]:
# Create the prediction using linear model

x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739],
        [0.5891]], grad_fn=<AddmmBackward>)


<!--Empty Space for separating topics-->


<h3>Práctico</h3>


Realice una predicción para el siguiente x usando lr.


In [37]:
# Practice: Use the linear regression model object lr to make the prediction.

x=torch.tensor([[1.0],[2.0],[3.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739],
        [0.5891],
        [1.1044]], grad_fn=<AddmmBackward>)


<!--Empty Space for separating topics-->


<h2 id="Cust">Construir modelos personalizados</h2>


Construyamos un módulo personalizado.


Importamos la siguiente librería:


In [38]:
# Library for this section

from torch import nn

Definimos la clase:


In [39]:
# Personalizar Linear Regression Class

class LR(nn.Module):
    
    # Constructor
    def __init__(self, input_size, output_size):
        # Hereda del padre
        super(LR, self).__init__()
        self.linear = nn.Linear(input_size, output_size)
    
    # Función de predicción
    def forward(self, x):
        out = self.linear(x)
        return out

Creamos un objeto usando el constructor. Imprimimos los parámetros y el modelo.


In [40]:
# Create the linear regression model. Print out the parameters.

lr = LR(1, 1)
print("The parameters: ", list(lr.parameters()))
print("Linear model: ", lr.linear)

The parameters:  [Parameter containing:
tensor([[-0.1939]], requires_grad=True), Parameter containing:
tensor([0.4694], requires_grad=True)]
Linear model:  Linear(in_features=1, out_features=1, bias=True)


<!--Empty Space for separating topics-->


Realicemos una predicción:

In [41]:
# Try our customize linear regression model with single input

x = torch.tensor([[1.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.2755]], grad_fn=<AddmmBackward>)


<!--Empty Space for separating topics-->


Ahora hagamos una predicción con múltiples muestras: 


In [42]:
# Try our customize linear regression model with multiple input

x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.2755],
        [0.0816]], grad_fn=<AddmmBackward>)


los parámetros son almacenados en un diccionario:

In [43]:
print("Python dictionary: ", lr.state_dict())
print("keys: ",lr.state_dict().keys())
print("values: ",lr.state_dict().values())


Python dictionary:  OrderedDict([('linear.weight', tensor([[-0.1939]])), ('linear.bias', tensor([0.4694]))])
keys:  odict_keys(['linear.weight', 'linear.bias'])
values:  odict_values([tensor([[-0.1939]]), tensor([0.4694])])


<!--Empty Space for separating topics-->


<h3>Práctica</h3>


Crear un objeto lr1 de la clase anterior y realizar una predicción usando el siguiente tensor x:


In [44]:
# Practice: Use the LR class to create a model and make a prediction of the following tensor.

x=torch.tensor([[1.0],[2.0],[3.0]])
lr1=LR(1,1)
yhat=lr(x)
yhat

tensor([[ 0.2755],
        [ 0.0816],
        [-0.1122]], grad_fn=<AddmmBackward>)